src/main/scala/com/daml/projection/JdbcAction.scala

```diff
@@ -162,42 +166,69 @@ object Sql {
+  final def apply(value: T, pos: Int): Bind.Applied = set(_, pos, value)
 }
 
-/** Sets a bind value on a PreparedStatement */
-@SerialVersionUID(1L)
-@FunctionalInterface
-trait Setter extends java.io.Serializable {
+/**
+ * Provides `Bind` values for [[ExecuteUpdate.bind[T](pos*]],[[ExecuteUpdate.bind[T](name*]], [[Binder.bind[T](pos*]]
+ * and [[Binder.bind[T](name*]] methods.
+ */
+object Bind {
+  type Used[-T] = Bind[_ >: T]
+
+  private def mk[T <: AnyVal](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = f(ps, pos, value)
+  }
+
+  private def mkN[T >: Null <: AnyRef](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = {
+      if (value eq null) ps.setNull(pos, java.sql.Types.NULL) else f(ps, pos, value)
+    }
+    // TODO SC would a specific thing from Types be better?
+  }
 
-  /** Sets a bind value on `ps` */
-  def set(ps: java.sql.PreparedStatement): Unit
+  implicit val Boolean: Bind[Boolean] = mk(_.setBoolean(_, _))
+  implicit val Byte: Bind[Byte] = mk(_.setByte(_, _))
+  implicit val Short: Bind[Short] = mk(_.setShort(_, _))
+  implicit val Int: Bind[Int] = mk(_.setInt(_, _))
+  implicit val Long: Bind[Long] = mk(_.setLong(_, _))
+  implicit val Float: Bind[Float] = mk(_.setFloat(_, _))
+  implicit val Double: Bind[Double] = mk(_.setDouble(_, _))
+  implicit val `scala BigDecimal`: Bind[BigDecimal] = mkN((ps, pos, x) => ps.setBigDecimal(pos, x.bigDecimal))
+  implicit val BigDecimal: Bind[java.math.BigDecimal] = mkN(_.setBigDecimal(_, _))
+  implicit val String: Bind[String] = mkN(_.setString(_, _))
+  implicit val Date: Bind[java.sql.Date] = mkN(_.setDate(_, _))
+  implicit val LocalDate: Bind[LocalDate] = mkN((ps, pos, x) => ps.setDate(pos, java.sql.Date.valueOf(x)))
+  implicit val Timestamp: Bind[java.sql.Timestamp] = mkN(_.setTimestamp(_, _))
+  implicit val LocalDateTime: Bind[LocalDateTime] =
+    mkN((ps, pos, x) => ps.setTimestamp(pos, java.sql.Timestamp.valueOf(x)))
+  implicit val Instant: Bind[Instant] =
+    mkN((ps, pos, x) => ps.setTimestamp(pos, new java.sql.Timestamp(x.toEpochMilli)))
+  implicit val Array: Bind[java.sql.Array] = mkN(_.setArray(_, _))
+  implicit val Null: Bind[Null] = mkN((ps, pos, _) => ps.setNull(pos, java.sql.Types.NULL))
+  // TODO SC is Types.NULL right?
+
+  implicit val Offset: Bind[Offset] = mkN((ps, pos, x) => ps.setString(pos, x.value))
+  implicit val ProjectionId: Bind[ProjectionId] = mkN((ps, pos, x) => ps.setString(pos, x.value))
+
+  implicit def Optional[A](implicit setter: Bind.Used[A]): Bind[Optional[A]] =
+    mkN((ps, pos, x) => x.ifPresentOrElse(setter.set(ps, pos, _), () => ps.setNull(pos, java.sql.Types.NULL)))
+  implicit def Option[A](implicit setter: Bind.Used[A]): Bind[Option[A]] =
+    mkN((ps, pos, x) => x.map(setter.set(ps, pos, _)).getOrElse(ps.setNull(pos, java.sql.Types.NULL)))
```

Comment:
This is the other null question: do we want to change the second argument to `setNull` here? If so, we should add the value to `Bind`, so these definitions can pull it from `setter`.

Reply:
From JDBC javadoc:

> Note: Not all databases allow for a non-typed Null to be sent to the backend. For maximum portability, the setNull or the setObject(String parameterName, Object x, int sqlType) method should be used instead of setObject(String parameterName, Object x).

I think we can do this on a follow up issue. Right now we only support postgres, and setting null works there as we have done here.

Reply:
https://github.com/digital-asset/daml/issues/15871